AvaTax.TaxModule.Data/Model/AvaLineItem.cs

```diff
@@ -19,26 +24,80 @@ public virtual LineItemModel FromTaxLine(TaxLine taxLine)
             return this;
         }
 
-        public virtual LineItemModel FromOrderLine(LineItem orderLine)
+        public virtual LineItemModel FromOrderLine(LineItem orderLine, CustomerOrder order, Store store, IFulfillmentCenterService fulfillmentCenterService)
         {
             number = orderLine.Id;
             itemCode = orderLine.Sku;
             description = orderLine.Name;
             taxCode = orderLine.TaxType;
             amount = orderLine.ExtendedPrice;
             quantity = orderLine.Quantity;
+
+            addresses = BuildAddressesModel(orderLine.FulfillmentCenterId, order, store, fulfillmentCenterService);
+
             return this;
         }
 
-        public virtual LineItemModel FromOrderShipment(Shipment shipment)
+        public virtual LineItemModel FromOrderShipment(Shipment shipment, CustomerOrder order, Store store, IFulfillmentCenterService fulfillmentCenterService)
         {
             number = shipment.Id;
             itemCode = shipment.Number;
             description = shipment.ShipmentMethodCode;
-            taxCode = shipment.TaxType;
             quantity = 1;
             amount = shipment.Sum;
+
```

Comment:
Also the convert logic shouldn't contains any complex logic such as determination of tax code based on   objects relationships.

Reply:
Well, I agree that the conversion code became quite complicated, but I don't see a solution yet:
* We can move the tax type and source address resolution to some helper services (something like `IShipmentTaxTypeResolver`), but it won't solve the problem of passing services to model converters - it will just move it one level deeper.
* We can use DI for `Ava*` classes to inject needed services there, but it just doesn't seem right.
* We can move line item creation outside of the `AvaCreateTransactionModel` class, but it would make the conversion code usage much more complicated.

Do you have any other ideas?